Arrow Key functionality and extracted some movement into functions

RiordanIX opened the pull request:
h, j, k, and l movement logic has been moved into functions so they can
be shared with the arrow keys. The arrow keys work pretty well, there
are just a few cases that need fine tuned (like pressing an arrow key
immediately after pressing escape, that doesn't quite work right).

There are some more escape sequences that can be implemented, like
page up, page down, home, end, etc. I will work on those later.

RiordanIX committed:
Makefile name change, preparing for escapesequences

RiordanIX committed:
Arrow keys now work approximately correctly

tomsmeding commented:
I really thought I'd implemented input better than `cin.get()`. In [termio](https://github.com/tomsmeding/termio/blob/master/termio.c#L398), and therefore all things after that, I used `select(2)` to get input with a timeout to correctly capture escape sequences etc.
Anyway, this seems fine; but since you have a TODO: open, I'll let you finish that before I merge :)

RiordanIX commented:
Sounds good, I may be able to get to it today actually.

RiordanIX committed:
Input read change and arrow movement in normal

A few more input options in normal mode, specifically arrow keys and
backspace. Reading keyboard-input is changed a bit to make certain
keys easier.

tomsmeding commented:
This looks like great work! Can't test right now, but I'll probably look at it tomorrow.
I'm privately busy with a fully rewritten version that has a way better class structure and is based on my termio library, making keyboard input a lot easier. But of course you can continue working on this thing if you want :p

RiordanIX commented:
Whenever I did it that way, exiting the program kept it in raw mode. It does need to be passed by reference when changing it.

Whenever we declare the termios, like on 131, it creates a copy of the current state of the terminal.  After changing the attributes, you do need to set the attribute by reference to update the current state. So in the disable and enable timeout functions, it gets the current state, modifies the two attributes, then makes that the state of the terminal.

When exiting the application, it resets the terminal to how it was originally, because we never actually changed tios_bak, just copies of it.

**Edit:** I read up on it some more and you're right.  It just so happened that editing still worked after switching the timeouts on the terminal.  Fortunately, this is a two line fix.  We don't really need to backup the slight editing of the terminals, so this should work fine.

RiordanIX committed:
Fixed terminal issues

Modifying the terminal should respond properly now. While changing the
timeout property of the terminal, it was resetting some of the other
edited properties, but now it should keep those the same.

tomsmeding commented:
I do not understand exactly what you're saying above, but that commit is exactly what I meant. :)
And are you aware my comment about the passing by reference is about something totally unrelated?

tomsmeding commented on io.cpp:
```diff
@@ -40,6 +40,32 @@ Colour ansiTable[8]={black,red,green,yellow,blue,magenta,cyan,white};
 unordered_map<string,string> tputcache;
+enum KEY_ACTION{
+			KEY_NULL = 0,       /* NULL */
+			CTRL_C = 3,         /* Ctrl-c */
```
Are you using these constants anywhere?

RiordanIX replied:
Many of them. They are some specialty keys that can get returned by the `editorReadKey` function. Not all are being used right now.

As for my previous comment, yes, I do realize now that the 'passing by reference' comment was about something different.  I don't remember why I added that comment in the first place, but I can remove it if you'd like.

tomsmeding commented:
My terminal emits different key codes for things like delete and home/end, so those don't work for me, but because they weren't implemented before in the first place (I believe), this is fine.
I looked through your changes and even though I didn't thoroughly check everything, stuff compiles and seems to work fine, so thumbs up!

If you could apply [this patch](https://gist.github.com/8f5e8950c54c73f8a6071141d77e32ab) and commit it to this PR, that would be awesome. Then I think this is ready to merge.

RiordanIX committed:
Makefile change and patch addition

Changed the Makefile because I upgraded my Lua version and it no
longer worked with updated lua. I noticed the regular expression
was way out of whack, so I simplified it and made it actually work
on any version of Lua >= 5.0.  The patch by tomsmeding was also
added.

RiordanIX commented:
Alright, I added the patch and made a minor change in the Makefile. I realized I was doing the regex all wrong and made it actually work for any version of lua >=5.0, at least on linux.
